---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# catGenes

<!-- badges: start -->
<!-- badges: end -->

The _catGenes_ package is intended to help researchers in phylogenetics and phylogenomics to build a fully concatenated or combined (non-interleaved) dataset by automatically comparing individual DNA alignments. The user can concatenate the multiple DNA matrices by either running the _catGenes_ functions in [R environment ](https://www.r-project.org/) or using an interactive [Shiny app](https://shiny.rstudio.com/).

## Installation

You can install the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("domingoscardoso/catGenes")
```
## Standard format for input DNA alignments

### _Formatting sequences with scientific names and accession identifiers_

The concatenating functions available in _catGenes_ work by comparing the scientific names across the individual DNA alignments, when the input matrices do 
not have species duplicated with multiple accessions. If the species are duplicated
then the concatenating comparison will consider both the scientific names and an associated unique identifier. As such, before loading the DNA alignments and runing _catGenes_ concatenating functions, make sure they have the sequences consistently formatted. Depending on the input DNA alignments, the matrices must be formatted
as following:

#### _1. When the DNA alignments have just a single sequence per species_

The species/sequences may be simply named as __Genus_species__, just like the example below. The generic name could even be abbreviated but always keeping separated from the specific epithet like __G_species__. Note that the sequences 
may also be labeled by including "cf" or "aff" between the genus and specific epithet, but they must be separated by an underscore.

![Example with no identifiers in the sequences](labelling_no_identifiers.png)


The species/sequences may also be named as __Genus_species_everythingelse__ or
__G_species_everythingelse__, that is, the scientific name is separated by the accession identifier (e.g. collector surname and associated collection number). The "everything else" means that the sequence here may have as many identifiers as desired, provided that at least the scientific name is consistently formatted. For example, in addition to format the sequences just like in the example
below, note that you could also format like __Vatairea_fusca_Cardoso2939_JX152598__.

![Example with identifiers in the sequences](labelling_with_identifiers_no_duplicated_species.png)

In brief, no matter the input DNA alignments have or not an associated identifier,
the species/sequences should be consistently named, either with the genus name abbreviated or not, and always separated from the identifier(s) by an underscore.


#### _2. When the DNA alignments have any species duplicated with multiple accessions_
The _catGenes_ package has specific functions to concatenate the DNA alignments 
when any species are duplicated with multiple accessions, i.e. the species is represented by multiple sequences generated from different individuals. In this case, the concatenation will consider both the consistently-formatted scientific names and their associated unique identifiers. Then, the sequ3nces/species must be
formatted as __Genus_species_identifier_everythingelse__ or
__G_species_identifier_everythingelse__, as exemplified below:

![Example when species are duplicated with multiple accessions](labelling_with_identifiers_and_duplicated_species.png)

### _Naming the individual DNA alignment files_
Before loading the individual DNA alignments for concatenation, we recommend that
each file is simply named with the corresponding gene names and do not use space 
or hyphen to separate the file name. For example: __ITS.nex__, __rbcL.nex__, __psbAtrnH.nex__, 
__COX1.nex__, etc. This is the best way to name the input files if you want to run _catGenes_ 
using the Shiny app. Also, it facilitates to load all DNA alignments by just using 
a _for_ loop that will create a list of the input genes ready for the concatenation.
Note that we load the DNA alignments by using [_ape_](https://www.rdocumentation.org/packages/ape/versions/5.4-1)'s function [read.nexus.data](https://www.rdocumentation.org/packages/ape/versions/5.4-1/topics/read.nexus.data).
See below an example using the available DNA alignments of the Vataireoid legumes 
that are stored within the _catGenes_ directory:

```{r, message=FALSE, warning=FALSE}
library(catGenes)

# Loading all individual DNA alignments for the concatenation
genes <- list.files(system.file("DNAlignments/Vataireoids", 
                                package = "catGenes"))
Vataireoids <- list()
for (i in genes){
  Vataireoids[[i]] <- ape::read.nexus.data(system.file("DNAlignments/Vataireoids", i, 
                                                       package = "catGenes"))
}
names(Vataireoids) <- gsub("[.].*", "", names(Vataireoids))
```

The created object __Vataireoids__ is a list of all input individual DNA alignments 
ready for the concatenation by using the available _catGenes_ functions. As 
you can see below, the containing input DNA alignments within the Vataireoids list
are named exactly how the files were originally named:
```{r}
names(Vataireoids)
```

Note that in your computer the individual DNA alignments could be loaded into a 
single list by adjusting the abovementioned code as follows:

```{r, eval=FALSE}
library(catGenes)

genes <- list.files("path_to_DNA_alignments_folder")
Vataireoids <- list()
for (i in genes){
  Vataireoids[[i]] <- ape::read.nexus.data(paste0("path_to_DNA_alignments_folder/", i))
}
names(Vataireoids) <- gsub("[.].*", "", names(Vataireoids))
```

If the input individual DNA alignments have any identifier associated
to the gene names, like __Vataireoids_ITS.nex__, __Vataireoids_matK.nex__, __Vataireoids_trnDT.nex__, etc., then you can load 
all such files by either importing each DNA alignment separately or by adjusting 
the abovementioned _for_ loop as follows:

```{r, eval=FALSE}
library(catGenes)

# Loading each DNA alignment from the working directory separately by using the
# ape's function read.nexus.data
ITS <- read.nexus.data("Vataireoids_ITS.nex")
matK <- read.nexus.data("Vataireoids_matK.nex")
trnDT <- read.nexus.data("Vataireoids_trnDT.nex")

# Loading just the Vataireoid DNA alignments even if there are other DNA alignments 
# in the same working directory
genes <- list.files("path_to_DNA_alignments_folder")
Vataireoids <- list()
for (i in genes[grepl("Vataireoids", genes)]){
  Vataireoids[[i]] <- ape::read.nexus.data(paste0("path_to_DNA_alignments_folder/", i))
}
# Deleting all characters between "Vataireoids_" and ".nex" so as to keep just the
# name of the genes in the vataireoids list of named DNA alignments
names(Vataireoids) <- gsub(".*_(.+)[.].*", "\\1", names(Vataireoids))
```



## Basic Usage

### _Loading the DNA matrices and running the concatenation_
```{r, message=FALSE, warning=FALSE}
library(catGenes)

# Loading all individual DNA alignments for the concatenation
genes <- list.files(system.file("DNAlignments/Vataireoids", 
                                package = "catGenes"))

# For simplicity, this example will include just three DNA matrices
Vataireoids <- list()
for (i in genes[1:3]){
  Vataireoids[[i]] <- ape::read.nexus.data(system.file("DNAlignments/Vataireoids", i, 
                                                       package = "catGenes"))
}
names(Vataireoids) <- gsub("[.].*", "", names(Vataireoids))

# Running the concatenating analysis
# Note that the main steps during concatenation will be printed in the R console
cat_datasets <- catfullGenes(Vataireoids,
                             shortaxlabel = TRUE,
                             missdata = TRUE)

```

### _Writing the concatenated matrix in nexus format_

This new function __writeNexus__ will write a nexus-formatted combined dataset with 
each individual gene alignment as interleaved and including a preliminary [MrBayes](http://nbisweden.github.io/MrBayes/) command block, including the charset
of each partition. Note that you may choose to concatenate the DNA alignments as non-interleaved and without the charset at the end of the matrix block.
```{r, message=FALSE, warning=FALSE}
writeNexus(cat_datasets, 
           file = "Vataireoids.nex",
           bayesblock = TRUE, 
           interleave = TRUE)
```

See below a screenshot of the beggining of the nexus-formatted concatenated matrix:

![catGenes Shiny app](concatenatedmatrix1.png)

See below a screenshot of the end of the nexus-formatted concatenated matrix, 
showing the charset of each partition:

![catGenes Shiny app](concatenatedmatrix2.png)

### _Writing the concatenated matrix in phylip format_
This new function __writePhylip__ will write both a phylip-formatted concatenated matrix and an associated partition file for [RAxML](https://cme.h-its.org/exelixis/web/software/raxml/index.html) concatenated phylogenetic analysis using a mixed/partitioned model, as available in [CIPRES](http://www.phylo.org/).
```{r, message=FALSE, warning=FALSE}
writePhylip(cat_datasets, 
            file = "Vataireoids_dataset.phy",
            catalignments = TRUE,
            partitionfile = TRUE)

```

See below a screenshot of the phylip-formatted concatenated matrix:

![catGenes Shiny app](concatenatedmatrix3.png)

See below a screenshot of the partition file to be uploaded in CIPRES for a RAxML 
concatenated phylogenetic analysis using a mixed/partitioned model:

![catGenes Shiny app](concatenatedmatrix3_partitition.png)

## Documentation
A detailed description of the _catGenes_' full functionality will soon be available
in [here](https://github.com/domingoscardoso/catGenes/tree/main/vignettes/).



## Shiny App
The _catGenes_ package also comes with a Shiny app that can be called by using the following command:
```{r, eval=FALSE}
run_catGenes()
```
Below some screenshots of the _catGenes_ application that will open automatically in the default internet browser, or the viewer if running from [RStudio](https://rstudio.com/).

![catGenes Shiny app](shiny1.png)

![catGenes Shiny app](shiny2.png)

![catGenes Shiny app; donwload button](shiny3.png)


## Citation
Cardoso, D., Cavalcante, Q. & Vilela, B. (2020). catGenes: a new R package for combining multiple DNA alignments for multigene analysis in phylogenetics and phylogenomics. https://github.com/domingoscardoso/catGenes
