---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# catGenes

<!-- badges: start -->
<!-- badges: end -->

The catGenes package is intended to help researchers in phylogenetics and phylogenomics to build a fully concatenated or combined (non-interleaved) dataset by automatically comparing individual DNA alignments.

## Installation

You can install the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("domingoscardoso/catGenes")
```
## Standard format for input DNA alignments

#### _Naming the individual DNA alignment files_
Before loading the individual DNA alignments for concatenation, we recommend that
each file is simply named with the corresponding gene names and do not use space 
or hyphen to separate the file name. For example: __ITS.nex__, __rbcL.nex__, __psbAtrnH.nex__, 
__COX1.nex__, etc. This is the best way to name the input files if you want to run _catGenes_ 
using the Shiny app. Also, it facilitates to load all DNA alignments by just using 
a for loop that will create a list of the input genes ready for the concatenation.
Note that we load the DNA alignments by using _ape_'s function __read.nexus.data__.
See below an example using the available DNA alignments of the Vataireoid legumes 
that are stored within the _catGenes_ directory:

```{r, message=FALSE, warning=FALSE}
library(catGenes)

# Loading all individual DNA alignments for the concatenation
genes <- list.files(system.file("DNAlignments/Vataireoids", 
                                package = "catGenes"))
Vataireoids <- list()
for (i in genes){
  Vataireoids[[i]] <- ape::read.nexus.data(system.file("DNAlignments/Vataireoids", i, 
                                                       package = "catGenes"))
}
names(Vataireoids) <- gsub("[.].*", "", names(Vataireoids))
```

The created object __Vataireoids__ is a list of all input individual DNA alignments 
ready for the concatenation by using the available _catGenes_ functions. As 
you can see below, the containing input DNA alignments within the Vataireoids list
are named exactly how the files were originally named:
```{r}
names(Vataireoids)
```

Note that in your computer the individual DNA alignments could be loaded into a 
single list by adjusting the abovementioned code as follows:

```{r, eval=FALSE}
library(catGenes)

genes <- list.files("path_to_DNA_alignments_folder")
Vataireoids <- list()
for (i in genes){
  Vataireoids[[i]] <- ape::read.nexus.data(paste0("path_to_DNA_alignments_folder/", i))
}
names(Vataireoids) <- gsub("[.].*", "", names(Vataireoids))
```

If the input individual DNA alignments have any identifier associated
to the gene names, like __Vataireoids_ITS.nex__, __Vataireoids_matK.nex__, __Vataireoids_trnDT.nex__, etc., then you can load 
all such files by either importing each DNA alignment separately or by adjusting 
the abovementioned for loop as follows:

```{r, eval=FALSE}
library(catGenes)

# Loading each DNA alignment from the working directory separately by using the
# ape's function read.nexus.data
ITS <- read.nexus.data("Vataireoids_ITS.nex")
matK <- read.nexus.data("Vataireoids_matK.nex")
trnDT <- read.nexus.data("Vataireoids_trnDT.nex")

# Loading just the Vataireoid DNA alignments even if there are other DNA alignments 
# in the same working directory
genes <- list.files("path_to_DNA_alignments_folder")
Vataireoids <- list()
for (i in genes[grepl("Vataireoids", genes)]){
  Vataireoids[[i]] <- ape::read.nexus.data(paste0("path_to_DNA_alignments_folder/", i))
}
# Deleting all characters between "Vataireoids_" and ".nex" so as to keep just the
# name of the genes in the vataireoids list of named DNA alignments
names(Vataireoids) <- gsub(".*_(.+)[.].*", "\\1", names(Vataireoids))
```


#### _Species labels and associated identifiers within the DNA alignments_

Before loading the DNA alignments to run catGenes functions, make sure the DNA aligments have the species names format as following:

(i) Underscores separate the genus and species name from the collector+number+genbank accession

Genus_species_everythingelse

(ii) so... no problem if the genus name is abbreviated

G_species_everythingelse  

```{r Luetzelburgia}
names(Luetzelburgia$ITS)
```

